From b9793da9aa03f0e99bd4d7de3c25639931c3bb4c Mon Sep 17 00:00:00 2001
From: Daniel Andersson <dandersson@chromium.org>
Date: Tue, 15 Feb 2022 21:03:09 +0000
Subject: [PATCH] Prevent UAF in aura::Window::CleanupGestureState.

Cancelling active touches may, under certain circumstances, end up
destroying the window that it was called from. This change adds code
to detect this and does an early exit if it happens.

Details and reproduction steps in the bug.

(cherry picked from commit 5a2a306942bc7e794ec147f685f5eb44e3746111)

Test: Manually verified.
Bug: 1292271
Change-Id: I48934b3e062111544ae0b2782ea1516739e6f851
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3456997
Reviewed-by: Xiaoqian Dai <xdai@chromium.org>
Reviewed-by: Avi Drissman <avi@chromium.org>
Commit-Queue: Daniel Andersson <dandersson@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#970825}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3465006
Auto-Submit: Daniel Andersson <dandersson@chromium.org>
Bot-Commit: Rubber Stamper <rubber-stamper@appspot.gserviceaccount.com>
Commit-Queue: Avi Drissman <avi@chromium.org>
Cr-Commit-Position: refs/branch-heads/4758@{#1166}
Cr-Branched-From: 4a2cf4baf90326df19c3ee70ff987960d59a386e-refs/heads/main@{#950365}
---

diff --git a/ui/aura/window.cc b/ui/aura/window.cc
index 4e3b40d..b667d59 100644
--- a/ui/aura/window.cc
+++ b/ui/aura/window.cc
@@ -9,7 +9,6 @@
 #include <algorithm>
 #include <utility>
 
-#include "base/auto_reset.h"
 #include "base/bind.h"
 #include "base/callback.h"
 #include "base/callback_helpers.h"
@@ -1252,11 +1251,18 @@
   // happen through some event handlers for CancelActiveTouches().
   if (cleaning_up_gesture_state_)
     return false;
+  cleaning_up_gesture_state_ = true;
 
-  base::AutoReset<bool> in_cleanup(&cleaning_up_gesture_state_, true);
+  // Cancelling active touches may end up destroying this window. We use a
+  // tracker to detect this.
+  // TODO(crbug.com/1292271): Add a regression test for this.
+  WindowTracker tracking_this({this});
+
   bool state_modified = false;
   Env* env = Env::GetInstance();
   state_modified |= env->gesture_recognizer()->CancelActiveTouches(this);
+  if (!tracking_this.Contains(this))
+    return state_modified;
   state_modified |= env->gesture_recognizer()->CleanupStateForConsumer(this);
   // Potentially event handlers for CancelActiveTouches() within
   // CleanupGestureState may change the window hierarchy (or reorder the
@@ -1267,6 +1273,7 @@
     Window* child = children.Pop();
     state_modified |= child->CleanupGestureState();
   }
+  cleaning_up_gesture_state_ = false;
   return state_modified;
 }
 
